### PR TITLE
[SPARK-35234][CORE] Reserve the format of stage failureMessage

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

`failureMessage` is already formatted, but `replaceAll("\n", " ")` destroyed the format. This PR fixed it.

### Why are the changes needed?

The formatted error message is easier to read and debug.

### Does this PR introduce _any_ user-facing change?

Yes, users see the clear error message in the application log.

(Note I changed a little bit to let the test throw exception intentionally. The test itself is good.)

Before: 
2141619490903_ pic_hd

After:

2151619490955_ pic_hd

### How was this patch tested?

Manually tested.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @mridulm @tgravescs @attilapiros

**Comment by attilapiros** [comment]
There is another case of using `replaceAll` on the message:

https://github.com/apache/spark/blob/78caf0a53e60d81e6211faadfc45fede5ef9c941/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1957

What about fixing that one, too?

**Comment by Ngone51** [comment]
SGTM

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1954]: Note that we don't append `\n` for  this "Most recent failure reason" because `message` already contains it:

```scala
val message = s"Stage failed because barrier task $task finished unsuccessfully.\n" +
  failure.toErrorString
```

**Comment by attilapiros** [comment]
@Ngone51 the images you attached to the PR description are the logs of a running test? 
If yes that would be not a user-facing change but a developer-facing one :) 

But I think these messages are landing on the UI and those should be attached to the PR description.

**Comment by Ngone51** [comment]
Yes, the screenshot is from a test. It's for convenient purpose. But the stage failure is shown to users directly, right? So I consider it a user-facing change.

Attach UI changes sounds like better idea, which is surely a user-facing change. I'll try.

**Comment by Ngone51** [comment]
@attilapiros I have done the experiment on UI. And it turns out that the format in UI is always correct. That's because we passed the formatted `failureMessage` to UI directly without any change. See:

https://github.com/apache/spark/blob/38ef4771d447f6135382ee2767b3f32b96cb1b0e/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L1732

**Comment by attilapiros** [comment]
The description would be perfect with the UI screenshot and it would be so good to take a look how the message is actually rendered on UI. 

@Ngone51 So I would like to kindly ask you to make those screenshots (before and after) by for example doing a temporarily code change!

**Comment by Ngone51** [comment]
@attilapiros I'm not sure if I misunderstood your request. I have done the UI experiment (https://github.com/apache/spark/pull/32356#issuecomment-827294571) and it turns out UI isn't affected. So I think I don't have to add the UI screenshot.

**Comment by attilapiros** [comment]
@Ngone51 Oh I see. Then please update the "Does this PR introduce any user-facing change?" section to "No."

As users cannot see / should not care about test logs.

**Comment by attilapiros** [comment]
In the title you can add [TESTS].

**Comment by Ngone51** [comment]
> As users cannot see / should not care about test logs.

@attilapiros It's not only able to see in the test, but also in a user application. Here, I used the test as an example to show the difference.

And probably it's my bad not to mention in advance: I changed a little bit in the test in order to show the error message (it's captured previously). So, I'm not fixing the test.

**Comment by attilapiros** [comment]
I think I see what you mean and why I am confused: I know you are not fixing a test but as you used test logs and said the UI is not affected I thought it has only relevance for test execution. But of course this fixes the application log as well.
 
So let's mention "application log" in the PR description to help the others who are reading the commit message after the merge.

**Review by attilapiros — APPROVED** [review]
LGTM

**Comment by tgravescs** [comment]
changes look fine to me

**Comment by attilapiros** [comment]
I leave it here for 2 more days to let the others to review it and if no issue comes up I'll merge it (assuming it's still passing CI and no review is in progress).

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Will leave it to you @attilapiros then :-)

**Comment by attilapiros** [comment]
@HyukjinKwon I am keeping my promise made here: https://github.com/apache/spark/pull/32180#issuecomment-820454967

**Review by mridulm — APPROVED** [review]

**Comment by Ngone51** [comment]
thanks all!

**Pull request closed without merging**
